src/middlewared/middlewared/logging/sentry_crash_reporting.py

```diff
@@ -0,0 +1,86 @@
+import os
+import logging
+
+import sentry_sdk
+
+from middlewared.utils import MIDDLEWARE_RUN_DIR, sw_info
+
+DISABLED_SENTINEL = '.crashreporting_disabled'
+DISABLED_KEY = 'CRASHREPORTING_DISABLED'
+MAX_STRING_LEN = 10240
+
+
+class SentryCrashReporting:
+    """Sentry class for reporting crash info to a remote endpoint"""
+    enabled_in_settings = False
+
+    def __init__(self):
+        self.logger = logging.getLogger('middlewared.logging.SentryCrashReporting')
+        self.init_sentry_sdk()
+
+    def init_sentry_sdk(self):
+        dsn = 'https://11101daa5d5643fba21020af71900475:d60cd246ba684afbadd479653de2c216@sentry.ixsystems.com/2'
+        query = '?timeout=3'
+        dsn = f'{dsn}{query}'
+        release = sw_info['fullname']
+        sentry_sdk.init(dsn=dsn, release=release, default_integrations=False)
+        sentry_sdk.utils.MAX_STRING_LENGTH = MAX_STRING_LEN
+
+    def is_disabled(self):
+        """
+        Sentry Crash reporting can be disabled via:
+            1. system_settings table in the db or
+            2. presence of sentifnel file (`DISABLED_SENTINEL`) or
+            3. environment variable (`DISABLED_KEY`)
+        """
+        if not self.enabled_in_settings:
+            # disabed in db
+            return True
+        elif os.path.exists(os.path.join(MIDDLEWARE_RUN_DIR if sw_info()['stable'] else '/data', DISABLED_SENTINEL)):
+            # disabled because of presence of sentinel file
+            return True
+        elif os.environ.get(DISABLED_KEY, False):
+            # disabled via environment variable
+            return True
+        elif os.stat(__file__).st_dev != os.stat('/').st_dev:
+            # really??
```

Comment:
Middleware mounted from an external source http://docs.middleware.ixsystems.net/middleware/development.html#substituting-the-middleware-code-in-an-already-built-image

Reply:
Heh, yeah I realized that and forgot to remove comment 😄